prow/cmd/initupload/main.go

```diff
@@ -0,0 +1,180 @@
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/option"
+
+	"k8s.io/test-infra/prow/pjutil"
+	"k8s.io/test-infra/prow/pod-utils/clone"
+	"k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+var (
+	cloneLog = flag.String("clone-log", "", "path to the output file for the cloning step")
+
+	subDir = flag.String("sub-dir", "", "Optional sub-directory of the job's path to which artifacts are uploaded")
+
+	pathStrategy = flag.String("path-strategy", pathStrategyExplicit, "how to encode org and repo into GCS paths")
+	defaultOrg   = flag.String("default-org", "", "optional default org for GCS path encoding")
+	defaultRepo  = flag.String("default-repo", "", "optional default repo for GCS path encoding")
+
+	gcsBucket          = flag.String("gcs-bucket", "", "GCS bucket to upload into")
+	gceCredentialsFile = flag.String("gcs-credentials-file", "", "file where Google Cloud authentication credentials are stored")
+	dryRun             = flag.Bool("dry-run", true, "do not interact with GCS")
+)
+
+type pathStrategyType string
+
+const (
+	pathStrategyLegacy   pathStrategyType = "legacy"
+	pathStrategyExplicit                  = "explicit"
+	pathStrategySingle                    = "single"
+)
+
+func main() {
+	flag.Parse()
+
+	if !*dryRun {
+		if *gcsBucket != "" {
+			logrus.Fatal("No GCS bucket specified")
+		}
+
+		if *gceCredentialsFile != "" {
+			logrus.Fatal("No GCE credentials specified")
+		}
+	}
+
+	if err := validatePathOptions(pathStrategy, defaultOrg, defaultRepo); err != nil {
+		logrus.Fatalf("Invalid path options: %v", err)
+	}
+
+	var builder gcs.RepoPathBuilder
+	switch pathStrategyType(*pathStrategy) {
+	case pathStrategyExplicit:
+		builder = gcs.NewExplicitRepoPathBuilder()
+	case pathStrategyLegacy:
+		builder = gcs.NewLegacyRepoPathBuilder(*defaultOrg, *defaultRepo)
+	case pathStrategySingle:
+		builder = gcs.NewSingleDefaultRepoPathBuilder(*defaultOrg, *defaultRepo)
+	}
+
+	spec, err := pjutil.ResolveSpecFromEnv()
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not resolve job spec")
+	}
+
+	gcsPath := gcs.PathForSpec(spec, builder)
+	if *subDir != "" {
+		gcsPath = path.Join(gcsPath, *subDir)
+	}
+
+	var cloneRecords []clone.Record
+	data, err := ioutil.ReadFile(*cloneLog)
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not read clone log")
+	}
+	if err = json.Unmarshal(data, &cloneRecords); err != nil {
+		logrus.WithError(err).Fatal("Could not unmarshal clone records")
+	}
+
+	failed := false
+	buildLog := bytes.Buffer{}
+	for _, record := range cloneRecords {
+		buildLog.WriteString(clone.FormatRecord(record))
+		failed = failed || record.Failed
+	}
+
+	uploadTargets := map[string]gcs.UploadFunc{
+		path.Join(gcsPath, "build-log.txt"):      gcs.DataUpload(&buildLog),
+		path.Join(gcsPath, "clone-log.txt"):      gcs.DataUpload(&buildLog),
+		path.Join(gcsPath, "clone-records.json"): gcs.FileUpload(*cloneLog),
```

Comment:
`gcs.FileUpload(*cloneLog)` should be used for `clone-log.txt` and `clone-records.json` should be generated from `cloneRecords`.

Reply:
`*cloneLog` is the location of the JSON on disk, and `cloneRecords` is formatted nicely into `&buildLog` so the way we have it now is:

 - `clone-log.txt`: formatted clone output
 - `clone-records.json`: JSON metadata about clone operations

This seems appropriate without changes? WDYT?

Reply:
Ok, that makes more sense. Do we need to upload `gcs.DataUpload(&buildLog)` twice though?

Reply:
If we upload the build log later (from the test sidecar) this build log will be overwritten and nobody will know, but the clone log formatted nicely will still be around if anyone wants to look. If we fail right here, we need the clone log to be the build log otherwise people will have nothing to look at in Gubernator. I'm open to only uploading the clone log as `build-log.txt` if we've failed

Reply:
Either seems totally reasonable to me. Might be worth adding a comment though if you keep it as is.

Reply:
Decided to change it.